### ISSUE TITLE
[HOLD for payment 2023-09-06] [$1000] Web -  Inconsistent error messages text on wrong magic code

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#22017]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
Action Performed:
1. Signout if you are logged in
2. Enter an email > continue
3. Enter wrong magic code
4. Login
5. Go to contact methods
6. Add an email > verify
7. Enter wrong magic code

## Expected Result:
same error message

## Actual Result:
 Inconsistent error messages text on wrong magic code

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.35.4
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/93399543/a0f6fb15-a5e1-40b6-8410-5b4769b4489d

https://github.com/Expensify/App/assets/93399543/9d39325a-dc6a-4e63-b348-29da8bf35cac

**Expensify/Expensify Issue URL:**
**Issue reported by:** @chiragxarora
**Slack conversation:**  https://expensify.slack.com/archives/C049HHMV9SM/p1688221719614829

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0150dd5082869ecc1c</li>
        <li>Upwork Job ID: 1675276504517607424</li>
        <li>Last Price Increase: 2023-07-22</li>
<li>Automatic offers: </li>
<ul>
<li>abdulrahuman5196 | Reviewer | 26274048</li>
<li>chiragxarora | Reporter | 26274049</li>
</ul></li>
    </ul>
</details>

**Comment by kbecciv** [comment]
## Proposal
### Please re-state the problem that we are trying to solve in this issue.
inconsistent error message on incorrect magic code
### What is the root cause of that problem?
Root cause of the issue is the different error messages being passed from backend. When user enters an incorrect code while signing in, different error message comes from backend, when wrong code on secondary login, different error message.
### What changes do you think we should make in order to solve the problem?
We should make the error message consistent on backend, we should pick one error message out of two and use it on both flows.
We should use the one which says "incorrect magic code, please try again or request a new one" because other message mentioning code expiry doesn't make much sense
<details>
  <summary>Results</summary>
</details>
### What alternative solutions did you explore? (Optional)
None

**Comment by MitchExpensify** [comment]
Agree this is expected: "incorrect magic code, please try again or request a new one"

**Comment by abdulrahuman5196** [comment]
@chiragxarora  Could you provide references for your proposal? https://github.com/Expensify/App/issues/22017#issuecomment-1615943468

From code I could see places where we are setting the magic code error locally. Gentle suggestion would be to provide references and more information in proposal, so that we don't have to assume certain parts of proposal.

**Comment by alitoshmatov** [comment]
I think backend should change the error message, because right now it is showing what backend is sending in error
https://github.com/Expensify/App/blob/c614ed94a61ff37c248828a33f0f147cdb0a2143/src/pages/settings/Profile/Contacts/ValidateCodeForm/BaseValidateCodeForm.js#L168

**Comment by MitchExpensify** [comment]
We have a fresh proposal for you to review @abdulrahuman5196

**Comment by abdulrahuman5196** [comment]
@MitchExpensify 

In the signin page we show `This magic code is invalid or has expired, please request a new code and try again` which is got from backend

In contact verification page we show `Failed to validate contact method with given magic code. Please request a new code and try again.` which is shown with local data.


Which one should be the right one to maintain consistency? Based on that we can make the change.

**Comment by MitchExpensify** [comment]
This is expected: "Incorrect magic code, please try again or request a new one"

**Comment by nishancx** [comment]
I think this has been fixed, error message while verifying contact method has been replaced by a more appropriate message.

<img width="377" alt="Screenshot 2023-07-11 at 08 43 33" src="https://github.com/Expensify/App/assets/52850000/2d0a5c6a-e612-4ac8-896e-4cf7997d14bf">

**Comment by abdulrahuman5196** [comment]
The text is changed but still inconsistent as per the OP. Correct me if wrong. @MitchExpensify

**Comment by MitchExpensify** [comment]
Can you tell when that new text was introduced @abdulrahuman5196 ? curious to see that

**Comment by MitchExpensify** [comment]
friendly bump @abdulrahuman5196

**Comment by MitchExpensify** [comment]
No overdue, Melvin!

**Comment by MitchExpensify** [comment]
Thank you! Yes I agree [this](https://github.com/Expensify/App/issues/22017#issuecomment-1616170602) is the expected text

**Comment by MitchExpensify** [comment]
Confirmed [this](https://github.com/Expensify/App/issues/22017#issuecomment-1616170602) is the expected text. As such, do any of the existing proposals sound good to you @abdulrahuman5196 ?

**Comment by abdulrahuman5196** [comment]
Reviewing this now

**Comment by abdulrahuman5196** [comment]
On @chiragxarora 's proposal here https://github.com/Expensify/App/issues/22017#issuecomment-1616463336
There is one major issue - currently backend error messages are not translated. I think that could have been the historical reason of showing local error in terms for secondary validation and not for signin (Because signin is not expected to be translated AFAIK). So If we relay on backend error, we won't be able to shown translated content in secondary verification page.

**Comment by abdulrahuman5196** [comment]
@HezekielT 's proposal here https://github.com/Expensify/App/issues/22017#issuecomment-1617841843 looks good and works well. Since we are already showing the local error message, we can update it to have consistency.

🎀 👀 🎀
C+ Reviewed

cc: @MitchExpensify

**Comment by cead22** [comment]
@MitchExpensify @abdulrahuman5196 sorry if this was already discussed, but should we perhaps figure out what the final copy should be before making the update? Otherwise somebody can change one of them and we'll be back to square one

I **briefly** looked at the PR that introduced this copy (cc @Beamanator @aldo-expensify) and I couldn't find a linked issue with the Waiting for Copy label -- again, I took a quick look, so I could've missed it.

**Comment by cead22** [comment]
I guess we can also move forward with the proposal, since part of the checklist is to make an issue with the Waiting for Copy label to finalize the copy, but since it sounds like this message was updated recently, I think we should go a step further than "it's inconsistent, therefore it's a bug". 

I actually find the copies to be appropriate for the different cases/flows

**Comment by MitchExpensify** [comment]
I figured [this was the expected text](https://github.com/Expensify/App/issues/22017#issuecomment-1616170602) seeing as its taken from the most prominent real estate (Sign in) so if it is wrong I would think it would have been noticed by now. Is that solid enough reasoning to move forward with or do we need to get a label on this do you think @cead22 ?

**Comment by cead22** [comment]
I think that's reasonable. However, I think if we simply update this using this reasoning, then nothing stops the next person from thinking **this text is wrong, it's the same as in the sign in page, but this is a different flow**, which is also reasonable. And then we'll just continue creating PRs to update this over and over

I think we should discuss in product or with marketing such that when the next person thinks this is wrong, and finds the last PR that updated it, they can also find the slack discussion where this was agreed on, which should either settle it, or they can re-start the discussion to change it again

**Comment by cead22** [comment]
@chiragxarora yes, but I may have missed the point you were trying to make. Can you perhaps clarify? I think [my previous comment](https://github.com/Expensify/App/issues/22017#issuecomment-1656287067) still stands

> * firstly, we should remove/replace the `errorField` from the failure data and let the onyx use the error message received from backend while validating code

I agree with this, but we need to figure out what to do with the other message first, and I'm proposing we discuss what we want these messages to be moving forward, instead of simply changing them to be consistent even though the error messages make sense for the both flows

**Comment by abdulrahuman5196** [comment]
Agree on the concern @cead22 . I am assuming @MitchExpensify confirmed for this question. https://github.com/Expensify/App/issues/22017#issuecomment-1627395700.

We could check again from product if the confirmation is correct to be safe @MitchExpensify ?

**Comment by cead22** [comment]
Ah yes, that makes sense, and I think it's in line with what I commented [here](https://github.com/Expensify/App/issues/22017#issuecomment-1658056326)

**Comment by abdulrahuman5196** [comment]
@cead22 Just checking back. I assume we are waiting on product comeback.

On another note. from your comment

>>firstly, we should remove/replace the errorField from the failure data and let the onyx use the error message received from backend while validating code

>I agree with this, but we need to figure out what to do with the other message first, and I'm proposing we discuss what we want these messages to be moving forward, instead of simply changing them to be consistent even though the error messages make sense for the both flows


I was not inclined to this because, backend error messages don't have translation as of now. It is not a problem for login page, but will be in secondary verification, if we remove current onyx error and rely on backend error message, we would regress the translation capability in secondary verification page.

If people unknowingly changing the text is the concern means, we could add a comment mentioning a similar text is used in login page and some information and we can say change both places if required.

**Comment by MitchExpensify** [comment]
@cead22 is better placed to help out with the technical considerations here 👍

**Comment by cead22** [comment]
@MitchExpensify we're waiting for you to lead [this product discussion](https://github.com/Expensify/App/issues/22017#issuecomment-1656287067), not on anything technical

**Comment by MitchExpensify** [comment]
Ah got it! Thanks @cead22 

This seems be essentially be a copy issue so I'm applying the `Waiting for Copy` label.

@LLPeckham - what do you think is the best copy for all cases whenever 

In the signin page we show "This magic code is invalid or has expired, please request a new code and try again"

In contact verification page we show "Failed to validate contact method with given magic code. Please request a new code and try again"

Both make sense but we need to decide which is best for both cases so we don't keep changing them endlessly. Thanks!

**Comment by LLPeckham** [comment]
Got it. @MitchExpensify - Question: would this error code also show up if someone entered a code just incorrectly? It seems like we basically want to decide on a blanket error message for any time a code is either invalid or entered incorrectly? LMK if that is what we're trying to figure out here, because if so, I agree we should just have one standard error message. Thoughts on something like: "Incorrect or invalid magic code. Please try again or request a new code." 

With "This magic code is invalid or has expired, please request a new code and try again." invalid = expired, but doesn't cover the case where it's just incorrect, so I like saying it's either entered wrong or invalid altogether and the next steps for the user is to either try again or request a new code. 

What do ya think?

**Comment by cead22** [comment]
> Both make sense but we need to decide which is best for both cases so we don't keep changing them endlessly. Thanks!

To clarify, having different messages for different flows -- specially if they make sense for their respective flow, imo, is also an option

**Comment by MitchExpensify** [comment]
> "Incorrect or invalid magic code. Please try again or request a new code."

This sounds good to me @LLPeckham !

**Comment by LLPeckham** [comment]
@cead22 got it. So either I'd recommend one message per flow, which is if I'm understanding correctly:

- Flow 1 - When a user is trying to validate a contact method and uses an invalid/incorrect code. So the message would be: `Failed to validate contact method with given magic code. Please try again or request a new code.`

- Flow 2 - When a user is at the sign in page and uses an invalid/incorrect code. So the message would be: "Incorrect or invalid magic code. Please try again or request a new code."

Or just use the blanket code for both: `Incorrect or invalid magic code. Please try again or request a new code.`

I'm open to both if you think it makes sense to have 2 separate messages for the different scenarios (cc @MitchExpensify @cead22), but that's the copy I'd recomend for either case!

**Comment by LLPeckham** [comment]
I think we should just go with one message for this error code, so I'd suggest sticking with: `Incorrect or invalid magic code. Please try again or request a new code.`

**Comment by cead22** [comment]
Thanks!

@abdulrahuman5196 I think we're ready to move forward, do you still think @HezekielT 's [proposal](https://github.com/Expensify/App/issues/22017#issuecomment-1617841843) is the one to go with?

**Comment by abdulrahuman5196** [comment]
Yeah. I still think, @HezekielT 's proposal here https://github.com/Expensify/App/issues/22017#issuecomment-1617841843 looks good and works well. We also need translation for the new string. 

We also need to update the backend error message separately on server. @cead22 

🎀 👀 🎀
C+ Reviewed

**Comment by abdulrahuman5196** [comment]
> I think we should just go with one message for this error code, so I'd suggest sticking with: `Incorrect or invalid magic code. Please try again or request a new code.`

@LLPeckham If possible could you also provide the required translation for the new string?

**Comment by MitchExpensify** [comment]
Asked for translation [here](https://expensify.slack.com/archives/C21FRDWCV/p1692400395583009), will report back

**Comment by LLPeckham** [comment]
Bumped the thread in Slack for the translation

**Comment by LLPeckham** [comment]
Translation here:  `Código mágico incorrecto o no válido. Inténtalo de nuevo o solicita otro código.`

**Comment by cead22** [comment]
> We also need to update the backend error message separately on server. @cead22

@abdulrahuman5196 I can take care of this part (or ask an internal engineering to do it). What's the API call that's returning the error message we need to update?

**Comment by cead22** [comment]
The proposal is complete because we only expect front end changes in the proposal, or at least that's my understanding. If you feel differently, I think you should feel encouraged to start a discussion about this in #expensfy-open-source, but I don't think it's worth changing this decision after having selected a proposal and having a PR submitted

**Comment by MitchExpensify** [comment]
**Payment summary:**

Reporting: @chiragxarora $250
C: @HezekielT $1500
C+: @abdulrahuman5196 $1500

**Comment by abdulrahuman5196** [comment]
>The PR that introduced the bug has been identified. Link to the PR:
The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:

Not a regression. We have made the error message consistent.

>Determine if we should create a regression test for this bug.
If we decide to create a regression test for the bug, please propose the regression test steps to ensure the same bug will not reach production again.

Regression test needs to be added. But can only be added after backend fix is also done. So we can add regression test part of that.

@MitchExpensify / @cead22  Could you kindly create separate issue for backend string update if required?

**Comment by MitchExpensify** [comment]
> @MitchExpensify / @cead22 Could you kindly create separate issue for backend string update if required?

Is this update required do you think?

**Comment by MitchExpensify** [comment]
Reporting: @chiragxarora $250 - Paid and contract ended
C: @HezekielT $1500 - Please accept the Upwork offer 👍 
C+: @abdulrahuman5196 $1500 - Paid and contract ended

**Comment by MitchExpensify** [comment]
Thanks @HezekielT , Paid and contract ended 👍

**Comment by cead22** [comment]
> Is this update required do you think? @cead22

Not really since the PR to fix the backend message is already merged and deployed. @abdulrahuman5196 given that can you  add the regression test now?

**Comment by abdulrahuman5196** [comment]
Regression test:

1. Go to login
2. Enter your email address and click continue.
3. Add incorrect magic codes in the magic code inputs.
4. Verify that an error messages that says `Incorrect or invalid magic code. Please try again or request a new code.` is displayed.
5. Now enter the correct magic code and login.
6. Navigate to Settings => Profile => Contact methods
7. Click on New contact method.
8. Enter your email and click on Add option.
9. From Contact methods options click on Please verify this contact method.
10. Add incorrect magic codes in the magic code inputs.
11. Verify that an error messages that says `Incorrect or invalid magic code. Please try again or request a new code.` is displayed.
12. Change the language to Spanish.
13. Repeat step 10 and 11.
14. Verify that an error messages that says `Código mágico incorrecto o no válido. Inténtalo de nuevo o solicita otro código.` is displayed in spanish.

@cead22 @MitchExpensify

**Comment by MitchExpensify** [comment]
Added the test request. Thanks all! Closing
